<a href="https://colab.research.google.com/github/javier-marti-isasi/03MAIR-Algoritmos-de-optimizacion/blob/main/Javier_Marti_Isasi_Seminario.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de optimización - Seminario

Nombre y apellidos: Javier Martí Isasi

Url Colab: https://colab.research.google.com/drive/19NBPIxHwxK8WmoybKPEv-prWi_0tpXp5

Url GitHub: https://github.com/javier-marti-isasi/03MAIR-Algoritmos-de-optimizacion

Problema: 3. Combinar cifras y operaciones

# Problema 3. Combinar cifras y operaciones. Enunciado

- El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.

- Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)

- Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una cantidad dada. Un ejemplo sería para obtener el 4:

  4+2-6/3*1 = 4

- Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las siguientes cuestiones:
  - ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?
  - ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?
- Nota: Es posible usar la función de python “eval” para evaluar una expresión:

### ¿Cuántas posibilidades hay sin tener en cuenta las restricciones?

Consideraremos como restricciones que no podemos repetir ninguna operación ni ningún número.

El número de operaciones sin tener en cuenta las restricciones es de: 

 $$4^4*9^5=15116544$$

Es decir, las resultantes de las 4 combinaciones de las 4 operaciones y las 5 combinaciones de los 9 números, permitiendo la repetición de números y operaciones. 

Lo comprobaremos recorriendo el conjunto de posibilidades posibles:


In [ ]:
import math

In [ ]:
valores_tot = []
count = 0

# Combinaciones de operadores posibles
for O1 in list({"+","-","*","/"}):
  for O2 in list({"+","-","*","/"}):
    for O3 in list({"+","-","*","/"}):
      for O4 in list({"+","-","*","/"}):

        # Combinaciones de números posibles
        for N1 in list({1,2,3,4,5,6,7,8,9}):
          for N2 in list({1,2,3,4,5,6,7,8,9}):
            for N3 in list({1,2,3,4,5,6,7,8,9}):
              for N4 in list({1,2,3,4,5,6,7,8,9}):
                for N5 in list({1,2,3,4,5,6,7,8,9}):

                  # Evaluación
                  str_num_op=str(N1)+O1+str(N2)+O2+str(N3)+O3+str(N4)+O4+str(N5)
                  value=eval(str_num_op)

                  if value%1==0 and value not in valores_tot:
                    valores_tot.append(value)
                  count+=1

valores_tot_sorted = sorted(valores_tot)

print("RESULTADOS SIN RESTRICCIONES")
# print("\nEl conjunto de valores posibles es:")
# for val in valores_tot_sorted:
#   print(int(val))

print("\nEl valor máximo es: %d" %(valores_tot_sorted[-1]))
print("El valor mínimo es: %d" %(valores_tot_sorted[0]))

print("\nEl número de operaciones totales es: %d" %(count))

RESULTADOS SIN RESTRICCIONES

El valor máximo es: 59049
El valor mínimo es: -6560

El número de operaciones totales es: 15116544


Efectivamente, el número de operaciones posibles sin tener en cuenta las restricciones es de 15116544.

### ¿Cuántas posibilidades hay teniendo en cuenta todas las restricciones?

El número de operaciones teniendo en cuenta las restricciones es de:

  $$4!*(9!/4!)=362880$$
  
Es decir, las resultantes de las 4 combinaciones de las 4 operaciones y las 5 combinaciones de los 9 números, descontando las combinaciones con números y operaciones repetidos.

* Las 4 combinaciones de las 4 operaciones se puede representar por: $4!$

* Las 5 combinaciones de los 9 números se puede representar por: $9!/4!$

Lo comprobaremos recorriendo el conjunto de posibilidades posibles:

In [ ]:
valores_tot = []
count = 0

# Combinaciones de operadores posibles
for O1 in list({"+","-","*","/"}):
  for O2 in list({"+","-","*","/"}-{O1}):
    for O3 in list({"+","-","*","/"}-{O1,O2}):
      for O4 in list({"+","-","*","/"}-{O1,O2,O3}):

        # Combinaciones de números posibles
        for N1 in list({1,2,3,4,5,6,7,8,9}):
          for N2 in list({1,2,3,4,5,6,7,8,9}-{N1}):
            for N3 in list({1,2,3,4,5,6,7,8,9}-{N1,N2}):
              for N4 in list({1,2,3,4,5,6,7,8,9}-{N1,N2,N3}):
                for N5 in list({1,2,3,4,5,6,7,8,9}-{N1,N2,N3,N4}):

                  # Evaluación
                  str_num_op=str(N1)+O1+str(N2)+O2+str(N3)+O3+str(N4)+O4+str(N5)
                  value=eval(str_num_op)

                  if value%1==0 and value not in valores_tot:
                    valores_tot.append(value)
                  count+=1

valores_tot_sorted = sorted(valores_tot)


print("RESULTADOS CON RESTRICCIONES")
# print("\nEl conjunto de valores posibles es:")
# for val in valores_tot_sorted:
#   print(val)

print("\nEl número de valores posibles es: %d" %(len(valores_tot_sorted)))

print("\nEl valor máximo es: %d" %(valores_tot_sorted[-1]))
print("El valor mínimo es: %d" %(valores_tot_sorted[0]))

print("\nEl número de operaciones totales es: %d" %(count))


RESULTADOS CON RESTRICCIONES

El número de valores posibles es: 147

El valor máximo es: 77
El valor mínimo es: -69

El número de operaciones totales es: 362880


### ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?

En la función anterior comprobamos que:
- El valor máximo es: 77
- El valor mínimo es: -69

### ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?

En la función anterior descubrimos que el rango de valores enteros posibles son los que se incluyen en el intervalo:


$$\{\space x \in \mathbb{Z} \space| \space x\in [-69, 77] \space\}$$

Un total de 147 valores posibles a los que podemos encontrar solución mediente el algoritmo de fuerza bruta y el de ramificación y poda desarrollado más adelante.

### ¿Cuál es la estructura de datos que mejor se adapta al problema?

He seguido dos estructuras distintas para el algoritmo de furza bruta y para el de ramificación y poda (desarrollado más adelante).

\
FUERZA BRUTA

Para el algoritmo por fuerza bruta, comencé pensando que la estructura de datos que mejor se adaptaba al problema sería la tupla, ya que por su naturaleza, garantiza que los canjuntos de datos albergados en ella son únicos.

En un principio ideé un algortimo por fuerza bruta que guardaba las combinaciones de operciones y de números en dos tuplas (una tupla para las combinaciones de operaciones y otra para las combinaciones de los números). Para evaluar el cálculo resultante de estas dos tuplas, se pasaban ambas tuplas como parámetros de entrada de una función y devolvía el resultado de la operación. 

Sin embargo, gracias al método eval() simplifiqué la evaluación del resultado de la operación; ya no almacené las combinaciones en las dos tuplas. Simplemente las combinaciones se almacenan en strings; suficiente para evaluar el resultado por medio del método eval().

\
RAMIFICACIÓN Y PODA

Para abordar el problema de ramificación y poda tuve clara la estructura de datos desde el principio. Las posibles combinaciones se van contruyendo por nodos que generan el árbol. Estos nodos se construyen mediante listas ya que permite, de forma muy práctica, acceder a cada uno de los elementos que la forman y editarla, modificando o añadiendos elementos. Además, también permite almacenar listas dentro de listas, lo que solucionaría el problema de generar la pila con los nodos a evaluar.



### ¿Cuál es la función objetivo?

La función objetivo viene definida por la combinación de operaciones fundamentales. Cada una de las 24 combinaciones posibles de operaciones dará lugar a una función objetivo determinada. 

Por ejemplo, la combinación de operaciones ("+", "-", "*", "/"), dará como resultado la siguiente función objetivo:

$$f(a, b, c, d, e)=a+b-c*d/e$$

### ¿Es un problema de maximización o minimización?

Es un problema de búsqueda. El objetivo del problema es hallar una combinación de operaciones y números que dé como resultado un valor de entrada.

### Diseña un algoritmo para resolver el problema por fuerza bruta

In [ ]:
def fuerza_bruta(num):
  combinaciones = []

  # Combinaciones de operadores posibles
  for O1 in list({"+","-","*","/"}):
    for O2 in list({"+","-","*","/"}-{O1}):
      for O3 in list({"+","-","*","/"}-{O1,O2}):
        for O4 in list({"+","-","*","/"}-{O1,O2,O3}):

          # Combinaciones de números posibles
          for N1 in list({1,2,3,4,5,6,7,8,9}):
            for N2 in list({1,2,3,4,5,6,7,8,9}-{N1}):
              for N3 in list({1,2,3,4,5,6,7,8,9}-{N1,N2}):
                for N4 in list({1,2,3,4,5,6,7,8,9}-{N1,N2,N3}):
                  for N5 in list({1,2,3,4,5,6,7,8,9}-{N1,N2,N3,N4}):
                    
                    # Evaluación
                    str_num_op=str(N1)+O1+str(N2)+O2+str(N3)+O3+str(N4)+O4+str(N5)
                    value=eval(str_num_op)

                    # Si la evaluación coincide con el valor buscado, devolvemos el resultado de la combinación
                    if value==num:
                      return str_num_op

  # Si hemos recorrido todas las combinaciones y no hemos encontrado la solución, devolvemos el resultado
  return "No se ha escontrado una solución."

In [ ]:
import time
start_time = time.time()

valor = 77 # Data entry del valor a buscar
solucion = fuerza_bruta(valor)

print("Para un valor de %s," %(valor))
print("la solución encontrada es:")
print(solucion)

print("\nTiempo transcurrido: %s segundos" % (time.time() - start_time))

Para un valor de 77,
la solución encontrada es:
7/1+8*9-2

Tiempo transcurrido: 0.21002483367919922 segundos


### Calcula la complejidad del algoritmo por fuerza bruta

Tal y como está diseñado el algoritmo por fuerza bruta, el número de iteraciones a realizar queda difinido por el propio algoritmo, independientemente de los datos de entrada. Está diseñado para encontrar una combinación de 4 operaciones y 5 números. El caso que requiere un mayor número de operaciones ya está establecido, no depende de los datos de entrada. 

Por ello, considerando n el dato de entrada, considero que el orden de complejidad que más se aproxima sería el de orden constante, O(1).

He realizado una aproximación al número de operaciones elementales que realiza el algoritmo por fuerza bruta en el peor de los casos, con un resultado de 7311768 OE. Este peor caso sería si no ha encontrado ninguna solución tras recorrer todas las posibles soluciones.

Con el siguiente código intento aproximarlo:

In [ ]:
def fuerza_bruta_oe(num):
  combinaciones = []
  oe = 2  #entrar en la función y definir la variable combinaciones


  # Combinaciones de operadores posibles

  oe+=1 #definir la lista de operadores
  for O1 in list({"+","-","*","/"}):
    oe+=2 #restar 1 operador a los valores + definir la lista de operadores
    for O2 in list({"+","-","*","/"}-{O1}):
      oe+=3 #restar 2 operadores a los valores + definir la lista de operadores
      for O3 in list({"+","-","*","/"}-{O1,O2}):
        oe+=4 #restar 3 operadores a los valores + definir la lista de operadores
        for O4 in list({"+","-","*","/"}-{O1,O2,O3}):


          # Combinaciones de números posibles

          oe+=1 #definir la lista de operadores
          for N1 in list({1,2,3,4,5,6,7,8,9}):
            oe+=2 #restar 1 operador a los valores + definir la lista de operadores
            for N2 in list({1,2,3,4,5,6,7,8,9}-{N1}):
              oe+=3 #restar 2 operadores a los valores + definir la lista de operadores
              for N3 in list({1,2,3,4,5,6,7,8,9}-{N1,N2}):
                oe+=4 #restar 3 operadores a los valores + definir la lista de operadores
                for N4 in list({1,2,3,4,5,6,7,8,9}-{N1,N2,N3}):
                  oe+=5 #restar 4 operadores a los valores + definir la lista de operadores
                  for N5 in list({1,2,3,4,5,6,7,8,9}-{N1,N2,N3,N4}):
                    
                    # Evaluación
                    str_num_op=str(N1)+O1+str(N2)+O2+str(N3)+O3+str(N4)+O4+str(N5)
                    oe+=14 #convertir 5 integers en strings + concatenar 9 strings + definir 1 variable
                    value=eval(str_num_op)
                    oe+=4 #calcular 4 sumas

                    oe+=1 #hacer 1 comparación
                    if value==num:
                      return oe
                      oe+=2 #definir 1 variable y hacer append

  oe+=1 #devolver el valor oe
  return oe

print("Número de operaciones elementeles: %d OE" %(fuerza_bruta_oe(100)))

Número de operaciones elementeles: 7311768 OE


El número de operaciones disminuye cuanto más rápido encuentra una solución. La experiencia nos dice que para valores intermedios en el conjunto de valores posibles, las operaciones necesarias para encontrar una solución se reducen.

Como indiqué anteriormente, el intervalo de posibles valores es de:

$$\{\space x \in \mathbb{Z} \space| \space x\in [-69, 77] \space\}$$

Por lo que se realizarán menos operaciones para valores próximos al 0 que los próximos a los extremos.

Por poner otro ejemplo, el número aproximado de operaciones realizadas para el valor 0 es de 68954 OE.

### Diseña un algoritmo que mejore la complejidad del algoritmo por fuerza bruta. Argumenta por qué crees que mejora el algoritmo por fuerza bruta

In [ ]:
import copy

def evaluacion(num,nodo):
  num_items = len(nodo)

  # Si la longitud del nodo es 9, evaluamos la solución final
  if num_items==9:
    joined_string = "".join(str(x) for x in nodo)
    result = eval(joined_string)
    if result==num:
      return "solucion_encontrada"
    else:
      return "no_solucion_final"

  # Si la longitud del nodo es 7 o 8, ramificaremos siempre
    # En caso del nodo con longitud 7 porque solo hay un hijo posible (solo resta una operación por asignar)
    # En caso del nodo con longitud 8 porque es el último antes de la evaluación final
  elif num_items==7 or num_items==8:
    return "ramifica"

  # Para el resto de casos analizaremos los nodos si ramificamos o podamos
  else:
    # Si no podemos escapar de un número decimal, podamos
    if num_items%2!=0:
      if check_decimal(nodo)==False:
        return "poda"
    
    # Si podemos escapar de un número decimal, evaluamos la cota superior
    cota_sup=CS(nodo)
    if num>cota_sup:
      return "poda"
    else:
      return "ramifica"


def check_decimal(nodo):
  # Devuelve True o False si podemos o no conseguir un número entero con el resto de elementos
  # Para favorecer el rendimiento, no consideramos la posibilidad de obtener un número entero mediante la * si aún falta por elegir
    # Sabemos que siempre podremos encontrar otras soluciones
    # Permite podar muchos nodos que, de no hacerlo, la mayoría se tendrían que podar más adelante

  if "/" in nodo:
    nodo_str = "".join(str(x) for x in nodo)
    value=eval(nodo_str)
    return value.is_integer()
  else:
    return True



def CS(nodo):
  # Devuelve la cota superior
  nodo_str = "".join(str(x) for x in nodo)

  # Para un nodo de longitud 1
  if len(nodo)==1:
    num = nodo[0]
    list_num_resto = list({1,2,3,4,5,6,7,8,9}-{nodo[0]})

    if num<8:
      op_1="+"
      num_2=9
      op_2="*"
      num_3=8
      op_3="-"
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-3]

    else:
      op_1="*"
      num_2=list_num_resto[-1]
      op_2="+"
      num_3=list_num_resto[-2]
      op_3="-"
      num_4=1
      op_4="/"
      num_5=list_num_resto[-3]
    
    cota_sup=eval(nodo_str+op_1+str(num_2)+op_2+str(num_3)+op_3+str(num_4)+op_4+str(num_5))


  # Para un nodo de longitud 2
  elif len(nodo)==2:
    op = nodo[1]
    list_num_resto = list({1,2,3,4,5,6,7,8,9}-{nodo[0]})

    if op=="+":
      num_2=list_num_resto[-1]
      op_2="*"
      num_3=list_num_resto[-2]
      op_3="-"
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-3]

    if op=="-":
      num_2=list_num_resto[0]
      op_2="/"
      num_3=list_num_resto[-3]
      op_3="+"
      num_4=list_num_resto[-1]
      op_4="*"
      num_5=list_num_resto[-2]

    if op=="*":
      num_2=list_num_resto[-1]
      op_2="+"
      num_3=list_num_resto[-2]
      op_3="-"
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-3]

    if op=="/":
      num_2=list_num_resto[0]
      op_2="+"
      num_3=list_num_resto[-1]
      op_3="*"
      num_4=list_num_resto[-2]
      op_4="-"
      num_5=list_num_resto[1]
    
    cota_sup=eval(nodo_str+str(num_2)+op_2+str(num_3)+op_3+str(num_4)+op_4+str(num_5))


  # Para un nodo de longitud 3
  elif len(nodo)==3:
    op = nodo[1]
    list_num_resto = list({1,2,3,4,5,6,7,8,9}-{nodo[0]-nodo[2]})
       
    if op=="+":
      op_2="*"
      num_3=list_num_resto[-1]
      op_3="-"
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-2]
      
    elif op=="-":
      op_2="/"
      num_3=list_num_resto[-3]
      op_3="+"
      num_4=list_num_resto[-1]
      op_4="*"
      num_5=list_num_resto[-2]

    elif op=="*":
      op_2="+"
      num_3=list_num_resto[-1]
      op_3="-"
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-2]

    elif op=="/":
      op_2="+"
      num_3=list_num_resto[-1]
      op_3="*"
      num_4=list_num_resto[-2]
      op_4="-"
      num_5=list_num_resto[0]

    cota_sup=eval(nodo_str+op_2+str(num_3)+op_3+str(num_4)+op_4+str(num_5))


  # Para un nodo de longitud 4
  elif len(nodo)==4:
    list_op = [nodo[1],nodo[3]]
    list_num_resto = list({1,2,3,4,5,6,7,8,9}-{nodo[0],nodo[2]})
      
    if list_op==["+","-"]:
      num_3=list_num_resto[1]
      op_3="/"
      num_4=list_num_resto[-1]
      op_4="*"
      num_5=list_num_resto[0]

    elif list_op==["-","+"]:
      num_3=list_num_resto[-2]
      op_3="*"
      num_4=list_num_resto[-1]
      op_4="/"
      num_5=list_num_resto[0]
      
    elif list_op==["+","*"]:
      num_3=list_num_resto[-1]
      op_3="-"
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-2]

    elif list_op==["+","/"]:
      num_3=list_num_resto[1]
      op_3="*"
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[0]
        
    elif list_op==["-","*"]:
      num_3=list_num_resto[0]
      op_3="/"
      num_4=list_num_resto[-1]
      op_4="+"
      num_5=list_num_resto[-2]
       
    elif list_op==["-","/"]:
      num_3=list_num_resto[-3]
      op_3="+"
      num_4=list_num_resto[-1]
      op_4="*"
      num_5=list_num_resto[-2]
        
    elif list_op==["*","+"]:
      num_3=list_num_resto[-1]
      op_3="-"
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-2]
        
    elif list_op==["*","-"]:
      num_3=list_num_resto[0]
      op_3="/"
      num_4=list_num_resto[-2]
      op_4="+"
      num_5=list_num_resto[-1]
      
    elif list_op==["*","/"]:
      num_3=list_num_resto[0]
      op_3="+"
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[1]
        
    elif list_op==["/","+"]:
      num_3=list_num_resto[-2]
      op_3="*"
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[0]
        
    elif list_op==["/","-"]:
      num_3=list_num_resto[0]
      op_3="+"
      num_4=list_num_resto[-1]
      op_4="*"
      num_5=list_num_resto[-2]
        
    elif list_op==["/","*"]:
      num_3=list_num_resto[-2]
      op_3="+"
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[0]

    cota_sup=eval(nodo_str+str(num_3)+op_3+str(num_4)+op_4+str(num_5))


  # Para un nodo de longitud 5
  elif len(nodo)==5:
    list_op = [nodo[1],nodo[3]]
    list_num_resto = list({1,2,3,4,5,6,7,8,9}-{nodo[0],nodo[2],nodo[4]})
      
    if list_op==["+","-"]:
      op_3="/"
      num_4=list_num_resto[-1]
      op_4="*"
      num_5=list_num_resto[0]

    elif list_op==["-","+"]:
      op_3="*"
      num_4=list_num_resto[-1]
      op_4="/"
      num_5=list_num_resto[0]
      
    elif list_op==["+","*"]:
      op_3="-"
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-1]

    elif list_op==["+","/"]:
      op_3="*"
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[0]
        
    elif list_op==["-","*"]:
      op_3="/"
      num_4=list_num_resto[-1]
      op_4="+"
      num_5=list_num_resto[-2]
       
    elif list_op==["-","/"]:
      op_3="+"
      num_4=list_num_resto[-1]
      op_4="*"
      num_5=list_num_resto[-2]
        
    elif list_op==["*","+"]:
      op_3="-"
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-1]
        
    elif list_op==["*","-"]:
      op_3="/"
      num_4=list_num_resto[-2]
      op_4="+"
      num_5=list_num_resto[-1]
      
    elif list_op==["*","/"]:
      op_3="+"
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[0]
        
    elif list_op==["/","+"]:
      op_3="*"
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[0]
        
    elif list_op==["/","-"]:
      op_3="+"
      num_4=list_num_resto[-1]
      op_4="*"
      num_5=list_num_resto[-2]
        
    elif list_op==["/","*"]:
      op_3="+"
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[0]

    cota_sup=eval(nodo_str+op_3+str(num_4)+op_4+str(num_5))


  # Para un nodo de longitud 6
  elif len(nodo)==6:
    list_op = [nodo[1],nodo[3],nodo[5]]
    list_num_resto = list({1,2,3,4,5,6,7,8,9}-{nodo[0],nodo[2],nodo[4]})

    if list_op==["+","-","*"]:
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-1]

    elif list_op==["+","-","/"]:
      num_4=list_num_resto[-1]
      op_4="*"
      num_5=list_num_resto[0]

    elif list_op==["-","+","*"]:
      num_4=list_num_resto[-1]
      op_4="/"
      num_5=list_num_resto[0]

    elif list_op==["-","+","/"]:
      num_4=list_num_resto[0]
      op_4="*"
      num_5=list_num_resto[-1]

    elif list_op==["+","*","-"]:
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-1]

    elif list_op==["+","*","/"]:
      num_4=list_num_resto[1]
      op_4="-"
      num_5=list_num_resto[0]

    elif list_op==["+","/","-"]:
      num_4=list_num_resto[0]
      op_4="*"
      num_5=list_num_resto[1]

    elif list_op==["+","/","*"]:
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[0]

    elif list_op==["-","*","+"]:
      num_4=list_num_resto[-1]
      op_4="/"
      num_5=list_num_resto[0]

    elif list_op==["-","*","/"]:
      num_4=list_num_resto[0]
      op_4="+"
      num_5=list_num_resto[-1]

    elif list_op==["-","/","+"]:
      num_4=list_num_resto[-1]
      op_4="*"
      num_5=list_num_resto[-2]

    elif list_op==["-","/","*"]:
      num_4=list_num_resto[0]
      op_4="+"
      num_5=list_num_resto[-1]

    elif list_op==["*","+","-"]:
      num_4=list_num_resto[0]
      op_4="/"
      num_5=list_num_resto[-1]

    elif list_op==["*","+","/"]:
      num_4=list_num_resto[1]
      op_4="-"
      num_5=list_num_resto[0]

    elif list_op==["*","-","+"]:
      num_4=list_num_resto[-1]
      op_4="/"
      num_5=list_num_resto[0]

    elif list_op==["*","-","/"]:
      num_4=list_num_resto[-2]
      op_4="+"
      num_5=list_num_resto[-1]

    elif list_op==["*","/","+"]:
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[0]

    elif list_op==["*","/","-"]:
      num_4=list_num_resto[0]
      op_4="+"
      num_5=list_num_resto[-1]

    elif list_op==["/","+","-"]:
      num_4=list_num_resto[0]
      op_4="*"
      num_5=list_num_resto[1]

    elif list_op==["/","+","*"]:
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[0]

    elif list_op==["/","-","*"]:
      num_4=list_num_resto[0]
      op_4="+"
      num_5=list_num_resto[-1]

    elif list_op==["/","-","+"]:
      num_4=list_num_resto[-1]
      op_4="*"
      num_5=list_num_resto[-2]

    elif list_op==["/","*","+"]:
      num_4=list_num_resto[-1]
      op_4="-"
      num_5=list_num_resto[0]

    elif list_op==["/","*","-"]:
      num_4=list_num_resto[0]
      op_4="+"
      num_5=list_num_resto[-1]

    cota_sup=eval(nodo_str+str(num_4)+op_4+str(num_5))
    
  return cota_sup



def creacion_primogenito(nodo_padre):
  # Se añade al nodo el primer valor de los disponibles
  num_ele = len(nodo_padre)

  if num_ele==0:
    new_ele = 1
  elif num_ele==1:
    new_ele = "+"
  elif num_ele==2:
    new_ele = list({1,2,3,4,5,6,7,8,9}-{nodo_padre[0]})[0]
  elif num_ele==3:
    new_ele = list({"+","-","*","/"}-{nodo_padre[1]})[0]
  elif num_ele==4:
    new_ele = list({1,2,3,4,5,6,7,8,9}-{nodo_padre[0],nodo_padre[2]})[0]
  elif num_ele==5:
    new_ele = list({"+","-","*","/"}-{nodo_padre[1],nodo_padre[3]})[0]
  elif num_ele==6:
    new_ele = list({1,2,3,4,5,6,7,8,9}-{nodo_padre[0],nodo_padre[2],nodo_padre[4]})[0]
  elif num_ele==7:
    new_ele = list({"+","-","*","/"}-{nodo_padre[1],nodo_padre[3],nodo_padre[5]})[0]
  elif num_ele==8:
    new_ele = list({1,2,3,4,5,6,7,8,9}-{nodo_padre[0],nodo_padre[2],nodo_padre[4],nodo_padre[6]})[0]

  nodo_padre.append(new_ele)
  return nodo_padre


def creacion_hermano(nodo):
# Se crea un nodo hermano a partir de un nodo dado
  # Si no se puede crear otro hermano, se devulve 0
  # Si se puede crear otro hermano, se devuelve el nodo

  # Si ya habíamos llegado a la última rama, no se generan más ramas
  if nodo ==0:
    return

  # Busca el siguiente nodo hermano del dado
  num_ele = len(nodo)


  if num_ele==1:
    if nodo[0]==9:
      return 0
    else:
      change_ele = nodo[0]+1

  elif num_ele==2:
    list_resto = list({"+","-","*","/"})
    ele = nodo[1]
    pos = list_resto.index(ele)
    if list_resto[-1]==ele:
      return 0
    else:
      change_ele = list_resto[pos+1]

  elif num_ele==3:
    #Se definen las opciones posibles de cambio de valor
    list_resto = list({1,2,3,4,5,6,7,8,9}-{nodo[0]})
    ele = nodo[2]
    # Si el valor a cambiar es el último de la lista de opciones posibles, no se podrán crear más hermanos, se devuelve 0
    if list_resto[-1]==ele:
      return 0
    # Si el valor a cambiar NO es el último de la lista de opciones posibles, se crea otro hermano con el siguiente elemento de los valores posibles
    else:
      pos = list_resto.index(ele)
      change_ele = list_resto[pos+1]

  elif num_ele==4:
    list_resto = list({"+","-","*","/"}-{nodo[1]})
    ele = nodo[3]
    pos = list_resto.index(ele)
    if list_resto[-1]==ele:
      return 0
    else:
      change_ele = list_resto[pos+1]

  elif num_ele==5:
    list_resto = list({1,2,3,4,5,6,7,8,9}-{nodo[0],nodo[2]})
    ele = nodo[4]
    if list_resto[-1]==ele:
      return 0
    else:
      pos = list_resto.index(ele)
      change_ele = list_resto[pos+1]

  elif num_ele==6:
    list_resto = list({"+","-","*","/"}-{nodo[1],nodo[3]})
    ele = nodo[5]
    pos = list_resto.index(ele)
    if list_resto[-1]==ele:
      return 0
    else:
      change_ele = list_resto[pos+1]

  elif num_ele==7:
    list_resto = list({1,2,3,4,5,6,7,8,9}-{nodo[0],nodo[2],nodo[4]})
    ele = nodo[6]
    pos = list_resto.index(ele)
    if list_resto[-1]==ele:
      return 0
    else:
      change_ele = list_resto[pos+1]

  elif num_ele==8:
      return 0

  elif num_ele==9:
    list_resto = list({1,2,3,4,5,6,7,8,9}-{nodo[0],nodo[2],nodo[4],nodo[6]})
    ele = nodo[8]
    pos = list_resto.index(ele)
    if list_resto[-1]==ele:
      return 0
    else:
      change_ele = list_resto[pos+1]

  nodo[num_ele-1] = change_ele

  return nodo



def ramificacion_poda(num):
  # COMPROBACIÓN INICIAL
  i = 0
  if num>77 or num<-69:
    return "No hay ninguna solución posible.", i

  # INICIALIZAMOS
  # Creación del primer primogénito
  nodo = creacion_primogenito([])
  LIST_NODO = []

  # Creación de primero hermanos a partir del primogénito
  # Mientras que la función creacion_hermano() no devulva 0, se siguen buscando hermanos y  se almacenan en LIST_NODO
  while nodo!=0:
    LIST_NODO.append(nodo)
    nodo=copy.deepcopy(nodo) # deepcopy para cambiar la referencia de nodo_hijo
    nodo = creacion_hermano(nodo)

  # COMENZAMOS RECORRIDO CON LOS NODOS INICIALIZADOS
  # Recorremos los nodos de LIST_NODO
  while( len(LIST_NODO) > 0):
    nodo = LIST_NODO[i]
    
    # EVALUAMOS
    evaluacion_result = evaluacion(num, nodo)

    # Si el nodo es una solución, devolvemos el resultado
    if evaluacion_result == "solucion_encontrada":
      return nodo, i
    
    # Si el resultado de la evaluación = "ramifica", ramificamos
    elif evaluacion_result == "ramifica":

      # Creación deL primogénito
      nodo_hijo = creacion_primogenito(nodo)
      LIST_NODO_HIJOS = []

      # Creación de hermanos, a partir del primogénito
      while nodo_hijo!=0:
        LIST_NODO_HIJOS.append(nodo_hijo)
        nodo_hijo=copy.deepcopy(nodo_hijo) # deepcopy para cambiar la referencia de nodo_hijo
        nodo_hijo = creacion_hermano(nodo_hijo)
    
      # Añadimos los hijos a LIST_NODO
      LIST_NODO.extend(LIST_NODO_HIJOS) 

    i+=1

In [ ]:
import time
start_time = time.time()

valor = 77

print("Para un valor de %s," %(valor))
print("la solución encontrada es:\n %s" %(ramificacion_poda(valor)[0]))

print("\nTiempo trascurrido:\n %s segundos" % (time.time() - start_time))

Para un valor de 77,
la solución encontrada es:
 [7, '+', 8, '*', 9, '-', 2, '/', 1]

Tiempo trascurrido:
 0.01878809928894043 segundos


El algoritmo de ramificación y poda supone una mejora en el rendiminento de la búsqueda de la solución ya que permite acotar la evaluación de combinaciones posibles. El algoritmo va generando un árbol donde sólo se exploran (ramifican) aquellos nodos con posibilidad de obtener la solución al porblema. Esto disminuye la cantidad de nodos a evaluar.

Puede verse el diseño del árbol en el siguiente enlace:
https://docs.google.com/presentation/d/1AUbsbJAgIAZ86mtdo0KZKzMmfHVvBPCHceoHLw1MxIE

(también está incluido al final del PDF)

Para llegar a la solución, se ramifica el árbol de tal forma que se va incremetando sucesivamente un número u operación en la solución buscada (según el elemento que toque siguiendo el orden natural de izquierda a derecha). Por cada nuevo elemento se hace una ramificación de los nodos del nivel superior no podados (teniendo en cuenta la limitación de no poder repetir elementos). 

Cuando toca como elemento la cuarta operación, no se evaluará la poda ya que solo habrá una rama restante. Directamente evaluaremos la solución final añadiendo la última operación y el último número. Como se muestra en el diseño del árbol, tendrá 8 ramificaciones, 5 por la elección de los 5 números y 3 por la elección de 3 operadores.

La decisión de podar o no un nodo queda definido por:
* La evaluación de si podemos llegar a una cierta cantidad estableciendo su cota superior
* La aparición de decimales

El principal beneficio del algoritmo se obtiene si quisiéramos descubrir todas las soluciones posibles, ya que el árbol se va ramificando por niveles. Cuando encontramos una solución, estamos más próximos de encontrar el resto.

Pondré el ejemplo que se percibe más claramente. Obtendremos todas las solucines posibles (mediante fuerza bruta y ramificación y poda) tomando como valor de entrada = 77. Calcularemos el tiempo de ejecución en cada caso:

In [ ]:
def fuerza_bruta_todas_soluciones(num):
  soluciones = []

  # Combinaciones de operadores posibles
  for O1 in list({"+","-","*","/"}):
    for O2 in list({"+","-","*","/"}-{O1}):
      for O3 in list({"+","-","*","/"}-{O1,O2}):
        for O4 in list({"+","-","*","/"}-{O1,O2,O3}):

          # Combinaciones de números posibles
          for N1 in list({1,2,3,4,5,6,7,8,9}):
            for N2 in list({1,2,3,4,5,6,7,8,9}-{N1}):
              for N3 in list({1,2,3,4,5,6,7,8,9}-{N1,N2}):
                for N4 in list({1,2,3,4,5,6,7,8,9}-{N1,N2,N3}):
                  for N5 in list({1,2,3,4,5,6,7,8,9}-{N1,N2,N3,N4}):
                    
                    # Evaluación
                    str_num_op=str(N1)+O1+str(N2)+O2+str(N3)+O3+str(N4)+O4+str(N5)
                    value=eval(str_num_op)

                    # Si la evaluación coincide con el valor buscado, almacenamos en la solución
                    if value==num:
                      soluciones.append(str_num_op)

  # Cuando hemos recorrido todas las combinaciones, devolvemos el resultado
  return soluciones


def ramificacion_poda_todas_soluciones(num):
  # COMPROBACIÓN INICIAL
  i = 0
  if num>77 or num<-69:
    return "No hay ninguna solución posible.", i

  soluciones=[]

  # INICIALIZAMOS
  # Creación del primer primogénito
  nodo = creacion_primogenito([])
  LIST_NODO = []

  # Creación de primero hermanos a partir del primogénito
  # Mientras que la función creacion_hermano() no devulva 0, se siguen buscando hermanos y  se almacenan en LIST_NODO
  while nodo!=0:
    LIST_NODO.append(nodo)
    nodo=copy.deepcopy(nodo) # deepcopy para cambiar la referencia de nodo_hijo
    nodo = creacion_hermano(nodo)


  # COMENZAMOS RECORRIDO CON LOS NODOS INICIALIZADOS
  # Recorremos los nodos de LIST_NODO
  while( len(LIST_NODO) > 0):

    #Cuando acabe todos los nodos, devuelve las soluciones
    try:
       nodo = LIST_NODO[i]
    except:
      return soluciones

    
    # EVALUAMOS
    evaluacion_result = evaluacion(num, nodo)

    # Si el nodo es una solución, almacenamos el resultado
    if evaluacion_result == "solucion_encontrada":
      soluciones.append(nodo)
    
    # Si el resultado de la evaluación = "ramifica", ramificamos
    elif evaluacion_result == "ramifica":

      # Creación deL primogénito
      nodo_hijo = creacion_primogenito(nodo)
      LIST_NODO_HIJOS = []

      # Creación de hermanos, a partir del primogénito
      while nodo_hijo!=0:
        LIST_NODO_HIJOS.append(nodo_hijo)
        nodo_hijo=copy.deepcopy(nodo_hijo) # deepcopy para cambiar la referencia de nodo_hijo
        nodo_hijo = creacion_hermano(nodo_hijo)
    
      # Añadimos los hijos a LIST_NODO
      LIST_NODO.extend(LIST_NODO_HIJOS) 

    i+=1

In [ ]:
valor = 77
print("Para un valor de %s," %(valor))

start_time = time.time()
print("\nFUERZA BRUTA")
print("  -  Las soluciones son: %s" %(fuerza_bruta_todas_soluciones(valor)))
print("  -  Tiempo trascurrido: %s segundos" % (time.time() - start_time))


start_time = time.time()
print("\nRAMIFICACIÓN Y PODA")
print("  -  Las soluciones son: %s" %(ramificacion_poda_todas_soluciones(valor)))
print("  -  Tiempo trascurrido: %s segundos" % (time.time() - start_time))

Para un valor de 77,

FUERZA BRUTA
  -  Las soluciones son: ['7/1+8*9-2', '7/1+9*8-2', '7/1-2+8*9', '7/1-2+9*8', '8/1*9+7-2', '9/1*8+7-2', '8/1*9-2+7', '9/1*8-2+7', '7+8/1*9-2', '7+9/1*8-2', '7+8*9/1-2', '7+9*8/1-2', '7+8*9-2/1', '7+8*9-4/2', '7+8*9-6/3', '7+9*8-2/1', '7+9*8-4/2', '7+9*8-6/3', '7-2/1+8*9', '7-2/1+9*8', '7-4/2+8*9', '7-4/2+9*8', '7-6/3+8*9', '7-6/3+9*8', '7-2+8/1*9', '7-2+9/1*8', '7-2+8*9/1', '7-2+9*8/1', '8*9/1+7-2', '9*8/1+7-2', '8*9/1-2+7', '9*8/1-2+7', '8*9+7/1-2', '9*8+7/1-2', '8*9+7-2/1', '8*9+7-4/2', '8*9+7-6/3', '9*8+7-2/1', '9*8+7-4/2', '9*8+7-6/3', '8*9-2/1+7', '8*9-4/2+7', '8*9-6/3+7', '9*8-2/1+7', '9*8-4/2+7', '9*8-6/3+7', '8*9-2+7/1', '9*8-2+7/1']
  -  Tiempo trascurrido: 2.880164861679077 segundos

RAMIFICACIÓN Y PODA
  -  Las soluciones son: [[7, '+', 8, '*', 9, '-', 2, '/', 1], [7, '+', 8, '*', 9, '-', 4, '/', 2], [7, '+', 8, '*', 9, '-', 6, '/', 3], [7, '+', 9, '*', 8, '-', 2, '/', 1], [7, '+', 9, '*', 8, '-', 4, '/', 2], [7, '+', 9, '*', 8, '-', 6, '/'

Se observa que para obtener todas las soluciones posibles, el algoritmo de ramificación y poda es unas 100 veces más rápido.

Sin embargo, para valores inferiores esta mejora en ramificación y poda se reduce. Como detallo en el último punto del trabajo, el algoritmo podría optimizarse mejorando la efectividad de la poda mediante la definición de cotas inferiores. 

### Calcula la complejidad del algoritmo

En los algoritmos de ramificación y poda, la complejidad depende del número de nodos recorridos, de la complejidad en el manejo de la LNV (lista de nodos vivos) y en el cálculo de las estimaciones de coste. 

Consideraremos el manejo de la LNV y el cálculo de las estimaciones de coste como constantes multimplicativas que, aunque interfieren en el número de operaciones elementales realizadas, no interfieren en el orden de complejidad del algoritmo. Por tanto, la complejidad queda definido por el número de nodos recorridos.


De forma similar a como sucedía con el algoritmo de fuerza bruta, el orden de complejidad no depende de los datos de entrada. El árbol que se construye está limitado por el diseño del algoritmo, siempre buscará una solución para una combinación de 5 números y 4 operaciones. El árbol completo (sin podar ningún nodo) será el mismo para cualquier valor de entrada. 

Por lo tanto, considerando n el dato de entrada, considero que el orden de complejidad que más se aproxima también sería el de orden constante, O(1).

Sin embargo, el número de operaciones elementales a realizar sí que dependerá del valor de entrada. El valor de entrada influirá en la poda, por lo que nos hará explorar un determinado número de nodos antes de encontrar la solución.

En el peor de los casos, la poda será completamente ineficiente, obligándonos a ramificar todos los nodos y evaluar cada uno de ellos. En este caso hipotético tendríamos que evaluar todas las combinaciones posibles hasta llegar a una solución como en fuerza bruta. Además, en el peor de los casos, la solución a encontrar sería única y coincidiría con la última en evaluar. Sin embargo, podemos afirmar que esto nunca ocurrirá. 

Por lo tanto, en el peor de los casos (que nunca podríamos experimentar), el algoritmo tendría que recorrer 395532 nodos como demuestro a continuación:




In [ ]:
def ramificacion_poda_num_nodos(num):
  # COMPROBACIÓN INICIAL
  i = 0

  # INICIALIZAMOS
  # Creación del primer primogénito
  nodo = creacion_primogenito([])
  LIST_NODO = []

  # Creación de primero hermanos a partir del primogénito
  # Mientras que la función creacion_hermano() no devulva 0, se siguen buscando hermanos y  se almacenan en LIST_NODO
  while nodo!=0:
    LIST_NODO.append(nodo)
    nodo=copy.deepcopy(nodo) # deepcopy para cambiar la referencia de nodo_hijo
    nodo = creacion_hermano(nodo)


  # COMENZAMOS RECORRIDO CON LOS NODOS INICIALIZADOS
  # Recorremos los nodos de LIST_NODO
  while( len(LIST_NODO) > 0):

    #Cuando acabe todos los nodos, devuelve el número de nodos recorrido
    try:
       nodo = LIST_NODO[i]
    except:
      return i
    
    # EVALUAMOS
    evaluacion_result = evaluacion(num, nodo)

    # Si el nodo es una solución, devolvemos el resultado
    if evaluacion_result == "solucion_encontrada":
      return nodo, i
    
    # Forzamos que también ramifique si el resultado es "poda"
    elif evaluacion_result == "ramifica" or evaluacion_result == "poda":

      # Creación deL primogénito
      nodo_hijo = creacion_primogenito(nodo)
      LIST_NODO_HIJOS = []

      # Creación de hermanos, a partir del primogénito
      while nodo_hijo!=0:
        LIST_NODO_HIJOS.append(nodo_hijo)
        nodo_hijo=copy.deepcopy(nodo_hijo) # deepcopy para cambiar la referencia de nodo_hijo
        nodo_hijo = creacion_hermano(nodo_hijo)
    
      # Añadimos los hijos a LIST_NODO
      LIST_NODO.extend(LIST_NODO_HIJOS) 

    i+=1

In [ ]:
print("Número de nodos recorridos: %d" %(ramificacion_poda_num_nodos(100)))

Número de nodos recorridos: 395532


La estimación del número de operaciones a realizar queda determinada por la efectividad de la poda. En el ejercicio, el criterio que se utiliza para podar se determina a partir de una cota superior. Es por ello que los valores más altos son más probables a ser podados. Esto queda reflejado en la evaluación del código:

    if num>cota_sup:
      return "poda"
    else:
      return "ramifica"

El segundo criterio que se emplea para podar es la aparición de decimales.

A continuación, representaré en una tabla el número de iteraciones realizadas por cada valor de entrada para hacernos una idea del número de operaciones elementales que realiza el algoritmo. Incluyo también como atributos la solución encontrada y el tiempo de ejecución.

In [ ]:
import pandas as pd

resultados = {}
frame = pd.DataFrame(columns=['Valor', 'Num. iteraciones', 'Solución', 'Tiempo (seg)'])

for valor in range(-70,79):
  start_time = time.time()
  ram_poda = ramificacion_poda(valor)
  time_seg = time.time() - start_time

  resultados['Valor'] = valor
  resultados['Num. iteraciones'] = ram_poda[1]
  resultados['Solución'] = ram_poda[0]
  resultados['Tiempo (seg)'] = time_seg

  frame.loc[len(frame)] = resultados

pd.set_option('display.max_rows', None)
display(frame)


Valor Num. iteraciones                          Solución  Tiempo (seg)
0     -70                0  No hay ninguna solución posible.      0.000008
1     -69           102903       [1, +, 4, /, 2, -, 8, *, 9]      3.200997
2     -68           104643       [1, +, 6, /, 2, -, 8, *, 9]      3.071031
3     -67           123219       [2, +, 3, /, 1, -, 8, *, 9]      3.326929
4     -66           124119       [2, +, 4, /, 1, -, 8, *, 9]      3.356768
5     -65           125019       [2, +, 5, /, 1, -, 8, *, 9]      3.393536
6     -64           125919       [2, +, 6, /, 1, -, 8, *, 9]      3.303370
7     -63           126879       [2, +, 7, /, 1, -, 8, *, 9]      3.312336
8     -62           150231       [3, +, 7, /, 1, -, 8, *, 9]      3.574343
9     -61           173703       [4, +, 7, /, 1, -, 8, *, 9]      3.710022
10    -60           102730       [1, +, 4, /, 2, -, 7, *, 9]      3.187167
11    -59           104422       [1, +, 6, /, 2, -, 7, *, 9]      3.069482
12    -58           106131       [1, +, 8, /, 2, -, 7, *, 9]      3.122528
13    -57           123742       [2, +, 4, /, 1, -, 7, *, 9]      3.358052
14    -56           124642       [2, +, 5, /, 1, -, 7, *, 9]      3.274541
15    -55           125542       [2, +, 6, /, 1, -, 7, *, 9]      3.258459
16    -54           148942       [3, +, 6, /, 1, -, 7, *, 9]      3.549554
17    -53           102561       [1, +, 4, /, 2, -, 7, *, 8]      3.090071
18    -52           104253       [1, +, 6, /, 2, -, 7, *, 8]      3.277007
19    -51           102461       [1, +, 4, /, 2, -, 6, *, 9]      3.174536
20    -50           123513       [2, +, 4, /, 1, -, 7, *, 8]      3.283483
21    -49           105886       [1, +, 8, /, 2, -, 6, *, 9]      3.182118
22    -48           123281       [2, +, 4, /, 1, -, 6, *, 9]      3.346978
23    -47           124181       [2, +, 5, /, 1, -, 6, *, 9]      3.298668
24    -46           147461       [3, +, 5, /, 1, -, 6, *, 9]      3.611658
25    -45           102220       [1, +, 4, /, 2, -, 6, *, 8]      3.078870
26    -44           106678       [1, +, 9, /, 3, -, 6, *, 8]      3.114029
27    -43           122236       [2, +, 3, /, 1, -, 6, *, 8]      3.337949
28    -42           101952       [1, +, 4, /, 2, -, 5, *, 9]      3.056740
29    -41           103601       [1, +, 6, /, 2, -, 5, *, 9]      3.085293
30    -40           105257       [1, +, 8, /, 2, -, 5, *, 9]      3.188112
31    -39           101691       [1, +, 4, /, 2, -, 6, *, 7]      3.057214
32    -38           106101       [1, +, 9, /, 3, -, 6, *, 7]      3.077341
33    -37           101591       [1, +, 4, /, 2, -, 5, *, 8]      3.153415
34    -36           103264       [1, +, 6, /, 2, -, 5, *, 8]      3.065248
35    -35           121451       [2, +, 3, /, 1, -, 5, *, 8]      3.203062
36    -34           122171       [2, +, 4, /, 1, -, 5, *, 8]      3.370550
37    -33           103056       [1, +, 6, /, 3, -, 4, *, 9]      3.064786
38    -32           101053       [1, +, 3, -, 8, /, 2, *, 9]      3.010868
39    -31           101857       [1, +, 4, -, 8, /, 2, *, 9]      3.132283
40    -30           102577       [1, +, 5, -, 8, /, 2, *, 9]      3.052819
41    -29           102623       [1, +, 6, /, 3, -, 4, *, 8]      3.045640
42    -28           102467       [1, +, 6, /, 2, -, 4, *, 8]      3.133527
43    -27           100605       [1, +, 4, /, 2, -, 5, *, 6]      3.018656
44    -26           105039       [1, +, 9, /, 3, -, 5, *, 6]      3.087947
45    -25           102286       [1, +, 6, /, 3, -, 4, *, 7]      2.995989
46    -24           100056       [1, +, 3, -, 7, /, 2, *, 8]      3.085487
47    -23            99853       [1, +, 3, -, 6, /, 2, *, 9]      2.971688
48    -22           100609       [1, +, 4, -, 6, /, 2, *, 9]      2.987925
49    -21            98869       [1, +, 2, -, 8, /, 3, *, 9]      3.041703
50    -20            99444       [1, +, 3, -, 6, /, 2, *, 8]      3.018034
51    -19           100248       [1, +, 4, -, 6, /, 2, *, 8]      2.966984
52    -18            98473   

Como observamos, tanto el número de iteraciones como el tiempo de ejecución, son inferiores con valores de entrada más altos.







### Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorio.

No le encuentro el sentido aplicarlo a este problema.

### Enumera las referencias que has utilizado (si ha sido necesario) para llevar a cabo el trabajo

- Introducción al diseño y análisis de algoritmos, R. C. T. Lee
- Departamento de lenguajes y sistemas informáticos (Univeridad de Sevilla): http://www.lsi.us.es/

### Describe brevemente en unas líneas como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño.

Como he comentado anteriormente, la efectividad del algoritmo depende en gran medida de la efectividad de la poda.

Para mejorar la efectividad de la poda, se podría:

* Definir una cota inferior que permitiera podar los nodos que no pueden alcanzar un valor inferior a esta cota.

* Incluir una función que evalúe los nodos más prometedores para evaluarlos en primer lugar y así encontrar una solución más rápido.

* Establecer como cotas superiores e inferiores números enteros

\
Por otro lado, una variación del problema podría ser que el número de combinaciones de operaciones y números sea variable. En el problema se determinó de partida que la solución tenía formarse por la combinación de 4 operaciones y 5 números. En caso de incrementar el número de operaciones, habría que establecer una nueva norma sobre la repetición de operaciones. 

Otra variación sería que la lista de combinaciones de entrada fuera también variable. En el problema se determinó que las operaciones posibles eran +, -, * y /. Podría incluirse la raíz cuadrada, exponentes, logaritmos... o incluso añadir paréntesis que determinen el orden de las operaciones. Respecto a las combinaciones de números, se podría ampliar o reducir el intervalo y permitir o no los números decimales y el 0.

En estos casos, donde las combinaciones de operaciones y números se incrementa, conyevaría a encontrar una función más genérica que determinara las cotas superiores e inferiores. Al aumentar la complejidad del problema tendríamos que acudir, probablemente, a técnicas heurísticas.